crypto/sm2/sm2_za.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2017 Ribose Inc. All Rights Reserved.
+ * Ported from Ribose contributions from Botan.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/sm2.h>
+#include <openssl/evp.h>
+#include <string.h>
+
+int SM2_compute_za(uint8_t *out,
+                   const EVP_MD *digest, const char *user_id, const EC_KEY *key)
+{
+    int rc = 0;
+
+    const EC_GROUP *group = EC_KEY_get0_group(key);
+
+    BN_CTX *ctx = NULL;         // ???
+
+    BIGNUM *p = BN_new();
+    BIGNUM *a = BN_new();
+    BIGNUM *b = BN_new();
+
+    BIGNUM *xG = BN_new();
+    BIGNUM *yG = BN_new();
+    BIGNUM *xA = BN_new();
+    BIGNUM *yA = BN_new();
+
+    int p_bytes = 0;
+    uint8_t *buf = NULL;
+
+    EVP_MD_CTX *hash = EVP_MD_CTX_new();
+
+    memset(out, 0, EVP_MD_size(digest));
+
+    if (EVP_DigestInit(hash, digest) == 0)
+        goto done;
+
+    /*
+       ZA=H256(ENTLA || IDA || a || b || xG || yG || xA || yA)
+     */
+
+    size_t uid_len = strlen(user_id);
```

Comment:
More declarations of variables in the middle of a code block, several below as well, the declarations need to move to the top.